src/cache.ts

```diff
@@ -20,306 +20,306 @@ const log = logging.createLogger('cache');
-          choices.set(name.substr(0, name.lastIndexOf('-')), valuestr.split(';'));
+    /**
+     * Read the contents of a CMakeCache.txt file.
+     * @param path Path to a CMakeCache.txt-format file
+     * @returns The CMake cache.
+     *
+     * @note The cache *may* not exist. In that case, the entries is empty and
+     * the `exists` property is `false`. Creating or modifying the file named by
+     * `path` has no effect on existing instance of this class.
+     */
+    static async fromPath(path: string): Promise<CMakeCache> {
+        log.debug(localize('reading.cmake.cache.file', 'Reading CMake cache file {0}', path));
+        const exists = await fs.exists(path);
+        if (exists) {
+            log.trace(localize('file.exists', 'File exists'));
+            const content = await fs.readFile(path);
+            log.trace(localize('file.contents.read.successfully', 'File contents read successfully'));
+            const entries = CMakeCache.parseCache(content.toString());
+            log.trace(localize('parsed.entries.from', 'Parsed {0} entries from {1}', entries.size, path));
+            return new CMakeCache(path, exists, entries);
         } else {
-          const key = name;
-          const typemap = {
-            BOOL: api.CacheEntryType.Bool,
-            STRING: api.CacheEntryType.String,
-            PATH: api.CacheEntryType.Path,
-            FILEPATH: api.CacheEntryType.FilePath,
-            INTERNAL: api.CacheEntryType.Internal,
-            UNINITIALIZED: api.CacheEntryType.Uninitialized,
-            STATIC: api.CacheEntryType.Static
-          } as {[type: string]: api.CacheEntryType | undefined};
-          const type = typemap[typename];
-          const docs = docs_acc.trim();
-          docs_acc = '';
-          if (type === undefined) {
-            rollbar.error(localize('cache.entry.unknown', 'Cache entry \'{0}\' has unknown type: \'{1}\'', name, typename));
-          } else {
-            log.trace(localize('constructing.new.cache.entry', 'Constructing a new cache entry from the given line'));
-            const entry = new Entry(key, valuestr, type, docs, false);
-            entry.serializedKey = serializedName;
-            entries.set(name, entry);
-          }
+            log.debug(localize('cache.file.does.not.exist', 'Cache file does not exist: Returning empty cache data'));
+            return new CMakeCache(path, exists, new Map());
+        }
+    }
+
+    /** Get a list of all cache entries */
+    get allEntries(): Entry[] {
+        return Array.from(this._entries.values());
+    }
+
+    /**
+     * Create a new instance. This is **private**. You may only create an instance
+     * via the `fromPath` static method.
+     * @param _path Path to the cache
+     * @param _exists Whether the file exists
+     * @param _entries Entries in the cache
+     */
+    private constructor(private readonly _path: string,
+        private readonly _exists: boolean,
```

Comment:
This is the kind of multi-line function signature spacing I was worried about introducing with a formatting change.

From the best I can tell, the TypeScript repo does not wrap lines in the function signatures even if they are long. I think we should follow their rules unless I can find a counter-example in their code.

Reply:
But maybe we just let this go for now until we get all the formatting configured.